source/compiler.ts

```diff
@@ -0,0 +1,174 @@
+  addJavaScript(info: FileData & FileInfo): void;
+  addJavaScript(info: FileInfo, finalize: () => Promise<FileData | null>): void;
+  error(error: { message: string }): void;
+  getArch(): string;
+  getContentsAsString(): string;
+  getPackageName(): string | undefined;
+  getPathInPackage(): string;
+  getSourceHash(): string;
+}
+
+type CompileResult = { data: string };
+
+class UniverseI18nCompiler extends CachingCompiler {
+  constructor() {
+    super({
+      compilerName: 'universe:i18n',
+      defaultCacheSize: 8 * 1024 * 1024,
+    });
+  }
+
+  addCompileResult(file: InputFile, compileResult: CompileResult) {
+    file.addJavaScript({ ...this.getFileInfo(file), ...compileResult });
+  }
+
+  compileOneFile(file: InputFile): CompileResult | null {
+    const sourcePath = file.getPathInPackage();
+    const source = file.getContentsAsString();
+    if (!source) {
+      return null;
+    }
+
+    const metadata = analyzePath(sourcePath);
+    const content = read(source, metadata.type);
+    if (content.error) {
+      file.error({ message: `Parsing Error: ${content.error.message}` });
+      return null;
+    }
+
+    const options = {
+      locale: getIfString(content.data, '_locale'),
+      namespace: getIfString(content.data, '_namespace'),
+      splitKey: getIfString(content.data, '_splitKey'),
+    };
+
+    const packageName = file.getPackageName();
+    const locale = i18n.normalize(metadata.locale ?? options.locale ?? '');
+    if (!locale) {
+      const extra = packageName ? ` in package: "${packageName}"` : '';
+      file.error({ message: `Cannot find locale for file "${file}"${extra}.` });
+      return null;
+    }
+
+    if (options.splitKey) {
+      splitKeys(content.data, options.splitKey);
+    }
+
+    return {
+      data: [
+        "Package['universe:i18n'].i18n.addTranslations(",
+        `'${locale}',`,
+        `'${options.namespace ?? packageName ?? ''}',`,
+        JSON.stringify(content.data),
+        ');',
+      ].join(''),
+    };
+  }
+
+  compileOneFileLater(
+    file: InputFile,
+    getCompileResult: () => Promise<CompileResult | null>,
+  ) {
+    file.addJavaScript(this.getFileInfo(file), getCompileResult);
+  }
+
+  compileResultSize(compileResult: CompileResult) {
+    return compileResult.data.length;
+  }
+
+  getCacheKey(file: InputFile) {
+    return `${file.getArch()}#${file.getSourceHash()}`;
+  }
+
+  getFileInfo(file: InputFile): FileInfo {
+    const sourcePath = file.getPathInPackage();
+    const path = `${sourcePath}.js`;
+    return { path, sourcePath };
+  }
+}
+
+function analyzePath(sourcePath: string) {
+  const type = path.extname(sourcePath);
+  const locale = path.basename(sourcePath, `.i18n${type}`);
+  return { locale, type };
+}
+
+function getIfString(data: JSON, key: string) {
+  if (!isJSONObject(data) || !(key in data)) {
+    return undefined;
+  }
+
+  const value = data[key];
+  delete data[key];
+  return typeof value === 'string' ? value : undefined;
+}
```

Comment:
I think the "get" function shouldn't delete anything. Can we change its name to something that describes better its behavior? I have in mind something like "shift" ("shiftString"), or "getIfStringAndRemove", but I don't really like them as much... Hope someone has a better idea.